server/src/sql-tools/from-database/readers/constraint.reader.ts

```diff
@@ -0,0 +1,144 @@
+import { sql } from 'kysely';
+import { DatabaseReader } from 'src/sql-tools/from-database/readers/type';
+import { DatabaseActionType, DatabaseConstraintType } from 'src/sql-tools/types';
+
+export const readConstraints: DatabaseReader = async (schema, db) => {
+  const constraints = await db
+    .selectFrom('pg_constraint')
+    .innerJoin('pg_namespace', 'pg_namespace.oid', 'pg_constraint.connamespace') // namespace
+    .innerJoin('pg_class as source_table', (join) =>
+      join.onRef('source_table.oid', '=', 'pg_constraint.conrelid').on('source_table.relkind', 'in', [
+        // ordinary table
+        sql.lit('r'),
+        // partitioned table
+        sql.lit('p'),
+        // foreign table
+        sql.lit('f'),
+      ]),
+    ) // table
+    .leftJoin('pg_class as reference_table', 'reference_table.oid', 'pg_constraint.confrelid') // reference table
+    .select((eb) => [
+      'pg_constraint.contype as constraint_type',
+      'pg_constraint.conname as constraint_name',
+      'source_table.relname as table_name',
+      'reference_table.relname as reference_table_name',
+      'pg_constraint.confupdtype as update_action',
+      'pg_constraint.confdeltype as delete_action',
+      // 'pg_constraint.oid as constraint_id',
+      eb
+        .selectFrom('pg_attribute')
+        // matching table for PK, FK, and UQ
+        .whereRef('pg_attribute.attrelid', '=', 'pg_constraint.conrelid')
+        .whereRef('pg_attribute.attnum', '=', sql`any("pg_constraint"."conkey")`)
+        .select((eb) => eb.fn<string[]>('json_agg', ['pg_attribute.attname']).as('column_name'))
+        .as('column_names'),
+      eb
+        .selectFrom('pg_attribute')
+        // matching foreign table for FK
+        .whereRef('pg_attribute.attrelid', '=', 'pg_constraint.confrelid')
+        .whereRef('pg_attribute.attnum', '=', sql`any("pg_constraint"."confkey")`)
+        .select((eb) => eb.fn<string[]>('json_agg', ['pg_attribute.attname']).as('column_name'))
+        .as('reference_column_names'),
+      eb.fn<string>('pg_get_constraintdef', ['pg_constraint.oid']).as('expression'),
+    ])
+    .where('pg_namespace.nspname', '=', schema.schemaName)
+    .execute();
+
+  for (const constraint of constraints) {
+    const table = schema.tables.find((table) => table.name === constraint.table_name);
```

Comment:
This is somewhat inefficient but I suppose it's fine because both sets won't ever be particularly big